Python/optimizer_analysis.c

```diff
@@ -1,50 +1,1316 @@
+/*
+ * This file contains the optimizer for CPython uops.
+ * It performs a traditional data-flow analysis[1] over the trace of uops.
+ * Using the information gained, it chooses to emit, or skip certain instructions
+ * if possible.
+ *
+ * [1] For information on data-flow analysis, please see page 27 onwards in
+ * https://ilyasergey.net/CS4212/_static/lectures/PLDI-Week-12-dataflow.pdf
+ * Credits to the courses UPenn Compilers (CIS 341) and NUS Compiler Design (CS4212).
+ *
+ * */
 #include "Python.h"
-#include "opcode.h"
 #include "pycore_interp.h"
 #include "pycore_opcode_metadata.h"
 #include "pycore_opcode_utils.h"
 #include "pycore_pystate.h"       // _PyInterpreterState_GET()
 #include "pycore_uop_metadata.h"
 #include "pycore_long.h"
 #include "cpython/optimizer.h"
+#include "pycore_optimizer.h"
+#include "pycore_object.h"
+#include "pycore_dict.h"
+#include "pycore_function.h"
+#include "pycore_uop_metadata.h"
+#include "pycore_uop_ids.h"
+#include "pycore_range.h"
+
+#include <stdarg.h>
 #include <stdbool.h>
 #include <stdint.h>
 #include <stddef.h>
-#include "pycore_optimizer.h"
+
+#define MAX_ABSTRACT_INTERP_SIZE 2048
+
+#define OVERALLOCATE_FACTOR 3
+
+#define PEEPHOLE_MAX_ATTEMPTS 5
+
+#ifdef Py_DEBUG
+    static const char *const DEBUG_ENV = "PYTHON_OPT_DEBUG";
+    #define DPRINTF(level, ...) \
+    if (lltrace >= (level)) { printf(__VA_ARGS__); }
```

Comment:
I propose to change the `DPRINTF` macro here to use a static inline function to compute `lltrace` from the env var. This would mean you can drop about 6 very distracting `#ifdef Py_DEBUG` sections from various functions.
```suggestion
    static inline int get_lltrace() {
        char *uop_debug = Py_GETENV(DEBUG_ENV);
        int lltrace = 0;
        if (uop_debug != NULL && *uop_debug >= '0') {
            lltrace = *uop_debug - '0';  // TODO: Parse an int and all that
        }
        return lltrace;
    }
    #define DPRINTF(level, ...) \
    if (get_lltrace() >= (level)) { printf(__VA_ARGS__); }
```